### PR TITLE
[ML] Add internal flag to flush api to control whether or not to flush indices

### PR DESCRIPTION
**Pull request by edsavage**
When datafeeds send flush requests they don't require the indices to be refreshed (and in fact may be detrimentally expensive in the case of small bucket sizes).

This change adds an (internal) flag to control whether or not a refresh is required when flushing.

### CODE REVIEW

**Comment by edsavage** [comment]
retest

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java [hunk 1]
```diff
@@ -88,6 +92,7 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeOptionalString(advanceTime);
             out.writeOptionalString(skipTime);
             out.writeBoolean(waitForNormalization);
+            out.writeBoolean(refreshRequired);
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java:95]: This needs BWC protection. Please read the JavaDocs in `TransportVersion.java` to see how to do it now.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java [hunk 2]
```diff
@@ -77,6 +80,7 @@ public Request(StreamInput in) throws IOException {
             advanceTime = in.readOptionalString();
             skipTime = in.readOptionalString();
             waitForNormalization = in.readBoolean();
+            refreshRequired = in.readBoolean();
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java:83]: Needs BWC protection.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/params/FlushJobParams.java [hunk 3]
```diff
@@ -93,6 +100,10 @@ public boolean isWaitForNormalization() {
         return waitForNormalization;
     }
 
+    public Boolean refreshRequired() {
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/params/FlushJobParams.java:103]:
```suggestion
    public boolean refreshRequired() {
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java [hunk 4]
```diff
@@ -187,6 +212,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             if (skipTime != null) {
                 builder.field(SKIP_TIME.getPreferredName(), skipTime);
             }
+            builder.field(REFRESH_REQUIRED.getPreferredName(), refreshRequired);
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FlushJobAction.java:215]: Similarly, it shouldn't be included in the JSON representation. (Like `waitForNormalization` is not mentioned in this method.)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/output/FlushAcknowledgement.java [hunk 5]
```diff
@@ -28,44 +29,57 @@ public class FlushAcknowledgement implements ToXContentObject, Writeable {
...
     public static final ConstructingObjectParser<FlushAcknowledgement, Void> PARSER = new ConstructingObjectParser<>(
         TYPE.getPreferredName(),
-        a -> new FlushAcknowledgement((String) a[0], (Long) a[1])
+        a -> new FlushAcknowledgement((String) a[0], (Long) a[1], (Boolean) a[2])
     );
 
     static {
         PARSER.declareString(ConstructingObjectParser.constructorArg(), ID);
         PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), LAST_FINALIZED_BUCKET_END);
+        PARSER.declareBoolean(ConstructingObjectParser.optionalConstructorArg(), REFRESH_REQUIRED);
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/process/autodetect/output/FlushAcknowledgement.java:42]: Here it _is_ correct to include the internal field in the parser, as this is for internal communications with our C++ code.

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**